### ISSUE TITLE
Random corrupt saved workspace leading to segfault 

### ISSUE THREAD

**Issue by vbraun** [gap-system/gap#501]
Saved workspaces are randomly (but rarely) corrupt; SaveWorkspace can also segfault sometimes. Seems to rely on memory layout and can be triggered by randomly fiddling with gap memory pool size and environment variables. The following script generally leads to at least one segfault before running to completion:
~~~~bash
#!/bin/sh
GAPDIR=/tmp/gap-test/gap4r8
GAP=$GAPDIR/bin/x86_64-pc-linux-gnu-gcc-default64/gap

for i in `seq 0 100`
do
    for MEM in `seq 5 20`
    do
        echo "MEM = ${MEM}000k"
        echo 'SaveWorkspace("workspace");' | $GAP -r -A -m ${MEM}000k -l $GAPDIR > /dev/null
    	echo 'Basis(GF(2)^1);' | $GAP -q -r -A -l $GAPDIR -L workspace
	if [ $? -ne 0 ]; then
            echo "Error - workspace corrupt?"
            exit 1
        fi
    	echo 'Irr(CharacterTable(GL(2,7)));' | $GAP -r -A -l $GAPDIR -L workspace | grep Error
    done

    echo "Setting env, loop $i"
    for j in `seq 0 10`
    do 
        newvar=`mktemp -u tmpenvXXXXXXXXXXXXX`
    	newval=`mktemp -u tmpvalueXXXXXXXXXXXXXXXXXXX`
        export "$newvar=$newval"
    done
done
~~~~
Partial output showing a segfault:
~~~~
MEM = 20000k
CanonicalBasis( ( GF(2)^1 ) )
Setting env, loop 55
MEM = 5000k
CanonicalBasis( ( GF(2)^1 ) )
MEM = 6000k
CanonicalBasis( ( GF(2)^1 ) )
MEM = 7000k
CanonicalBasis( ( GF(2)^1 ) )
MEM = 8000k
CanonicalBasis( ( GF(2)^1 ) )
MEM = 9000k
./makeenv: line 11: 13719 Done                    echo 'SaveWorkspace("workspace");'
     13720 Segmentation fault      (core dumped) | $GAP -r -A -m ${MEM}000k -l $GAPDIR > /dev/null
~~~~
This is present in Gap-4.8.1 but has been there since at least 4.5.7 (issue ported from redmine http://tracker.gap-system.org/issues/224)

**Comment by ChrisJefferson** [comment]
I believe I've tracked this down. Actually, it's not a workspace bug (you can see this by doing something like the following in the script (still doing the environment variables / memory limit loop).

``` l := Maximum(List([10..20], x -> Group([(1,x),(2,x),(3,x)])));;  Irr(CharacterTable(GL(2,7)));;```

Hopefully fix incoming shortly, after I've run a bunch of tests to double-check I've really fixed it (as it is a hard to trigger bug!)

**Comment by alex-konovalov** [comment]
@ChrisJefferson I've tried your example `l := Maximum(List([10..20], x -> Group([(1,x),(2,x),(3,x)])));; Irr(CharacterTable(GL(2,7)));;` 100 times in a loop for several times, and all worked - extra care should be used for testing that the fix in #510 actually works...

**Comment by ChrisJefferson** [comment]
@alex-konovalov : Have you tried running the script from #510 ? You need the repeated changes in memory and adding environment variables to trigger this.

**Comment by alex-konovalov** [comment]
No - can't see any scripts attached to #510. But if it is a typo and you mean the script above, I haven't run it either.

**Comment by alex-konovalov** [comment]
For reference: initial PR #510, seen above is closed, had simply been replaced by #519 in order to appear in stable-4.8 branch. #519 had been merged into stable-4.8 and then propagated to master via further merges.
